helm-v3/006-repositories.md

```diff
@@ -1,40 +1,178 @@
-- A new `helm push` verb will be added for pushing charts to a repository
-- The index file will be JSON instead of YAML
-
-> This document is in-progress and needs input from Chart Museum.
-
-## Pushing Charts To A Repository
-
-Several registries have emerged with support for Helm charts including
-ChartMuseum and Quay (via Application Registries) along with extras built on
-top of other systems, such as object storage (e.g., see the [S3
-Plugin](https://github.com/hypnoglow/helm-s3)). These registries are asking for
-and could benefit from a capability to push packages from the Helm client.
-
-To support registries there will be a `helm push` command that works in a
-similar manner to `helm fetch`. It will support different systems via the
-URI/URL scheme and provide a default implementation for HTTP(S). Other systems,
-such as S3, can implement an uploader via a plugin for schemes Helm does not
-provide an implementation for.
-
-> [name=Matt Farina] There very well may be additional work here around
-> registry search and pluggable or otherwise extendable authenication
-> mechanisms.
-> 
-> [name=Adnan Abdulhussein] What would the default HTTP(S) implementation be?
-> [name=Matt Fisher] Adnan, would you mind clarifying your comment a little?
-> 
-> [name=Adnan Abdulhussein] Sorry, the proposal mentions there will be a
-> default implemention for `helm push` to a regular HTTP(S) chart repository.
-> It's unclear to me how this would actually work.
-> 
-> [name=Adnan Abdulhussein] Sounds like this is something that will be fleshed
-> out more in Josh's proposal, so I'll wait for that :)
+- A repository v2 specification will be defined, and the index file will be
+JSON (`index.json` vs `index.yaml`)
+
+## Capabilities API
+
+The Capabilties API is designed to provide Helm with information such as:
+
+1. Does this domain support actions such as login or push?
+2. What supported method should the client use to perform a specific action?
+3. What metadata does the client need to initiate the action?
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
```

Comment:
side-note: once we have a clearer idea on some of the most basic capabilities, it might be useful to list them out in the documentation. Probably not valuable in a proposal but just a thought I wanted to share.